### PR TITLE
add from_existing_graph to neo4j vector

### PR DESCRIPTION
**Pull request by tomasonjo**
This PR adds the option to create a Neo4jvector instance from existing graph, which embeds existing text in the database and creates relevant indices.

Todo:

* [x] Docs
* [x] Tests

### CODE REVIEW

File: libs/langchain/langchain/vectorstores/neo4j_vector.py [hunk 1]
```diff
@@ -782,6 +788,108 @@ def from_documents(
             **kwargs,
         )
 
+    @classmethod
+    def from_existing_graph(
+        cls: Type[Neo4jVector],
+        embedding: Embeddings,
+        node_label: str,
+        embedding_node_property: str,
+        text_node_properties: List[str],
```
Comment (jexp) [inline libs/langchain/langchain/vectorstores/neo4j_vector.py:797]: should validate that the list is not empty

File: libs/langchain/langchain/vectorstores/neo4j_vector.py [hunk 2]
```diff
@@ -782,6 +788,108 @@ def from_documents(
...
+        Return Neo4jVector initialized from existing graph.
+        Neo4j credentials are required in the form of `url`, `username`,
+        and `password` and optional `database` parameters.
+        """
+
+        # Prefer retrieval query from params, otherwise construct it
+        if not retrieval_query:
+            retrieval_query = (
+                f"RETURN reduce(str='', k IN {text_node_properties} |"
+                " str + '\\n' + k + ':' + coalesce(node[k], '')) AS text, "
```
Comment (jexp) [inline libs/langchain/langchain/vectorstores/neo4j_vector.py:815]: add a space after the colon

File: libs/langchain/langchain/vectorstores/neo4j_vector.py [hunk 3]
```diff
@@ -782,6 +788,108 @@ def from_documents(
...
+            else:  # Validate that FTS and Vector index use the same information
+                if not fts_node_label == store.node_label:
+                    raise ValueError(
+                        "Vector and keyword index don't index the same node label"
+                    )
+
+        # Populate embeddings
+        while True:
+            fetch_query = (
+                f"MATCH (n:{node_label}) "
```
Comment (jexp) [inline libs/langchain/langchain/vectorstores/neo4j_vector.py:863]: quote with backticks?

File: libs/langchain/langchain/vectorstores/neo4j_vector.py [hunk 4]
```diff
@@ -782,6 +788,108 @@ def from_documents(
...
+                f"WHERE n.{embedding_node_property} IS null "
+                "AND any(k in $props WHERE n[k] IS NOT null) "
+                f"RETURN id(n) AS id, reduce("
+                "str='', k IN $props | str + '\\n' + k + ':' + n[k]) AS text "
```
Comment (jexp) [inline libs/langchain/langchain/vectorstores/neo4j_vector.py:867]: coalesce(n[k]) see the retrieval query above, 
can we reuse the expression
space after colon

we can also filter the non-null ones?

File: libs/langchain/langchain/vectorstores/neo4j_vector.py [hunk 5]
```diff
@@ -782,6 +788,108 @@ def from_documents(
...
+                        "Vector and keyword index don't index the same node label"
+                    )
+
+        # Populate embeddings
+        while True:
+            fetch_query = (
+                f"MATCH (n:{node_label}) "
+                f"WHERE n.{embedding_node_property} IS null "
+                "AND any(k in $props WHERE n[k] IS NOT null) "
+                f"RETURN id(n) AS id, reduce("
```
Comment (jexp) [inline libs/langchain/langchain/vectorstores/neo4j_vector.py:866]: elementId(n)

File: libs/langchain/langchain/vectorstores/neo4j_vector.py [hunk 6]
```diff
@@ -782,6 +788,108 @@ def from_documents(
...
+            params = {
+                "data": [
+                    {"id": el["id"], "embedding": embedding}
+                    for el, embedding in zip(data, text_embeddings)
+                ]
+            }
+
+            store.query(
+                "UNWIND $data AS row "
+                f"MATCH (n:{node_label}) "
```
Comment (jexp) [inline libs/langchain/langchain/vectorstores/neo4j_vector.py:882]: misses the `WHERE elementId(n) = row.id

File: libs/langchain/langchain/vectorstores/neo4j_vector.py [hunk 7]
```diff
@@ -782,6 +788,108 @@ def from_documents(
...
+                f"CALL db.create.setVectorProperty(n, "
+                f"'{embedding_node_property}', row.embedding) "
+                "YIELD node RETURN distinct 'done'",
```
Comment (jexp) [inline libs/langchain/langchain/vectorstores/neo4j_vector.py:885]: RETURN count(*) is nicer

**Review by jexp — CHANGES_REQUESTED** [review]
LGTM added a few comments, the most important one is fixing the update query

**Comment by tomasonjo** [comment]
Updated given the proposals by @jexp , and added tests and docs, so it should be good to go now.

**Comment by baskaryan** [comment]
thanks @tomasonjo @jexp!

**Pull request merged**